src/rv/comm/rcssserver/GameState.java

```diff
@@ -324,6 +350,66 @@ public void parse(SExp exp, WorldModel world) {
         }
     }
 
+    private void updateServerSpeed(long msgTime, float lastGameTime) {
+        // final long TIME_WINDOW_NS = 5000000000l;
+        final long TIME_WINDOW_MS = 5000;
+        final float DEFAULT_MSG_DELTA_S = 0.04f;
+
+        // Add message time info to map
+        if (serverMsgDeltas.isEmpty()) {
+            serverMsgDeltas.put(msgTime, -1.0f);
+            accumulatedServerTime_S = 0;
+        } else {
+            long lastMsgTime = serverMsgDeltas.lastKey();
+
+            float serverTimeDelta_S;
+            if (time - lastGameTime > 0) {
+                // We have a game time change for the amount of time passed
+                serverTimeDelta_S = time - lastGameTime;
+            } else {
+                // The game is paused so use DEFAULT_MSG_DELTA_S for amount of time passed
+                serverTimeDelta_S = DEFAULT_MSG_DELTA_S;
+            }
+
+            if (msgTime - lastMsgTime > 0) {
+                serverMsgDeltas.put(msgTime, serverTimeDelta_S + accumulatedServerTime_S);
+                accumulatedServerTime_S = 0;
+            } else {
+                // Messages are coming in so fast that they have the same time stamp so just save
+                // the time delta to add to the next entry with a new time stamp
+                accumulatedServerTime_S += serverTimeDelta_S;
+            }
+        }
+
+        // Remove map entries outside of time window
+        // SortedMap<Long, Float> oldEntries = serverMsgDeltas.headMap(msgTime - TIME_WINDOW_NS);
+        SortedMap<Long, Float> oldEntries = serverMsgDeltas.headMap(msgTime - TIME_WINDOW_MS);
+        while (!oldEntries.isEmpty()) {
+            serverMsgDeltas.remove(oldEntries.firstKey());
+        }
+
+        float sumDeltas = 0;
+
+        Float[] deltas = serverMsgDeltas.values().toArray(new Float[0]);
+        for (int i = 1; i < deltas.length; i++) {
+            float delta = deltas[i].floatValue();
```

Comment:
This call isn't necessary according to IntelliJ:

![](http://i.imgur.com/zKaVGd5.png)